test/Spec/BotPlutusInterface/Balance.hs

```diff
@@ -105,3 +165,131 @@ addUtxosForChange = do
+      initState =
+        def & utxos .~ [(txOutRef6, scrTxOut), (txOutRef7, usrTxOut)]
+          & contractEnv .~ contractEnv'
+      pabConf :: PABConfig
+      pabConf = def {pcOwnPubKeyHash = pkh3}
+      -- contractEnv' :: ContractEnvironment ()
+      contractEnv' = def {cePABConfig = pabConf}
+
+      -- Input UTxOs:
+      -- UTxO 1:
+      -- - From: User
+      -- - Amt : 1.001 ADA
+      -- UTxO 2:
+      -- - From: Script
+      -- - Amt : 0.5 ADA + 200 Tokens
+      --
+      -- Output UTxOs:
+      -- UTxO 1:
+      -- - To : User
+      -- - Amt: 1 ADA
+      -- UTxO 2:
+      -- - To : Script
+      -- - Amt: 1.0005 Ada + 200 Token
+      --
+      -- Fees   : 400 Lovelace
+      -- Change : 100 Lovelace
+
+      scrLkups =
+        Constraints.unspentOutputs (Map.fromList [(txOutRef6, scrTxOut'), (txOutRef7, usrTxOut')])
+          <> Constraints.ownPaymentPubKeyHash paymentPkh3
+      txConsts =
+        -- Pay the same datum to the script, but with more ada.
+        Constraints.mustPayToOtherScript valHash scrDatum (scrValue <> Ada.lovelaceValueOf 500)
+          -- <> Constraints.mustPayToOtherScript valHash scrDatum (Ada.lovelaceValueOf 1_000_000)
+          <> Constraints.mustPayToPubKey paymentPkh3 (Ada.lovelaceValueOf 1_000_000)
+          <> Constraints.mustSpendScriptOutput txOutRef6 Ledger.unitRedeemer
+          <> Constraints.mustSpendPubKeyOutput txOutRef7
+      eunbalancedTx = Constraints.mkTx @Void scrLkups txConsts
+
+  case eunbalancedTx of
+    Left mkTxErr -> assertFailure ("MkTx Error: " <> show mkTxErr)
+    Right unbalancedTx -> do
+      let (eRslt, _finalState) = runPABEffectPure initState (balanceTxIO @() @'[PABEffect ()] pabConf pkh3 unbalancedTx)
+      case eRslt of
+        (Left txt) -> assertFailure ("PAB effect error: " <> Text.unpack txt)
+        (Right (Left txt)) -> assertFailure $ "Balancing error: " <> Text.unpack txt -- <> "\n(Tx: " <> show unbalancedTx <> ")"
+        (Right (Right trx)) -> do
+          let scrTxOut'' = scrTxOut' & Ledger.ciTxOutValue <>~ Ada.lovelaceValueOf 500
+              scrTxOutNew = Ledger.toTxOut scrTxOut''
+          assertBool
+            ( "Expected UTxO not in output Tx."
+                <> "\nExpected UTxO: "
+                <> show scrTxOutNew
+                <> "\nNew UTxOs: "
+                <> show (txOutputs trx)
+                <> "\nUnbalanced UTxOs: "
+                <> show (txOutputs (unbalancedTx ^. OffChain.tx))
+            )
+            (scrTxOutNew `elem` txOutputs trx)
+
+-- Like the first one, but
+-- only has inputs from the script.
+dontAddChangeToDatum2 :: Assertion
+dontAddChangeToDatum2 = do
+  let scrTxOut' =
+        ScriptChainIndexTxOut
+          valAddr
+          (Right validator) -- (valHash, Just validator)
+          (Right scrDatum) -- (scrDatumHash, Just scrDatum)
+          (scrValue <> Ada.lovelaceValueOf 1_500_000)
+      scrTxOut = Ledger.toTxOut scrTxOut'
+      -- initState :: MockContractState ()
```

Comment:
Can be removed, probably. Or un-commented. And `contractEnv'` below too.

Reply:
Yeah, I left them as comments since I wasn't sure what their exact signature was, but wanted to make it clear what the overall type was. Since it turns out the types were just `SomeType ()`, I made them just normal type signatures.